### DISCUSSION TITLE
The Maintainer’s Lifecycle

### DISCUSSION THREAD

**Discussion by carltongibson** [carltongibson/notes#30] [Being an open source maintainer]
Hi all. I've put up a new post on _The Maintainer’s Lifecycle_, based on some of the discussions on #29. 

It's more questions than answers, deliberately so 🥳 — but it's something I think is important. 

Go check it out: [The Maintainer’s Lifecycle](https://noumenal.es/notes/open-source/maintainers-lifecycle/)

**Comment by nanorepublica** [comment]
It sounds like a similar question I failed to ask when I started my business/freelancing. What does the end look like for X? (where X is a package in this case)
Having an end in mind at least gives the project some direction in what it will and won't accept as features, how it might architect itself etc.

**Comment by kevinrenskers** [comment]
The biggest problem is when you're no longer using your own open source package and nobody is interested in talking over maintenance - even when you have tens of contributors. For me https://github.com/raml2html/raml2html comes to mind. Plenty of contributors over time, over 1100 stars, but nobody who wants to take ownership, and I stopped using RAML years ago. 

> Once we’re confortable [sic] as maintainers, do we start to see ourselves handing that off?

So absolutely yes, but there's gotta be people who want to take it on. For me the question is: what do you do what that doesn't happen? Just let the project die a slow death?

**Reply by carltongibson** [reply]
> For me the question is: what do you do what that doesn't happen? 

I think there's a whole load of things that can take some of the burden off, without needing people to (immediately) commit to taking on the package. 

For a Django package, it's often updating the CI and supported versions metadata for each release of Django. 
Documenting that, and then asking if someone can help with **that**, for **this release**, is a lot less intimidating that "Do you want to take over this package?" (And then who knows where that leads…)

> Just let the project die a slow death?

That's one option, yeah. 


Good spot on the typo. Fixed. Thanks.

**Reply by tim-schilling** [reply]
For me, this is why you need to build up the camaraderie and mentorship within your community. If the lead(ers) are at the point of wanting to hand things off, it's too far along to get someone [new] ready. Ideally, it should be immediately obvious who is going to replace the existing leadership when that time comes. That said, mentoring and building community are significantly different skills and tasks than the software skills of maintainership.

Maybe a low effort starting point is to ask someone to lead a project / release early on. Or something else equally difficult / gnarly that's more than just "reproduce bug, fix, create PR, cleanup, merge".

**Reply by bckohan** [reply]
It should definitely be easier to do the baseline maintenance for mature packages which usually seem to involve 4 kinds of tasks:

1. Updating CI matrix and testing against evolving upstream dependencies
2. Minor tweaks to fix usage of deprecated standard lib or dependency interfaces
3. Issuing new releases when CI passes against new upstream dependency versions
4. Fixing bugs introduced by new features in upstream dependencies that conflict with structural assumptions of your project.

It seems to me that 1 and 3 could be automated, 2 should be fast and low barrier work - great for new contributors. 4 is the only one that may require unfortunate amounts of time, but given how mature Django is, should be quite rare.

[django-polymorphic](https://github.com/jazzband/django-polymorphic) has been stuck in a 1-2-3 workflow for more than a year now even though [someone is willing to fix it](https://github.com/jazzband/django-polymorphic/issues/598).

What are the challenges here (aside from the obvious human resource one)? The way I see it:

1. **Trust**. I pucker when I think about giving [Someone From The Internet](https://www.akamai.com/blog/security-research/critical-linux-backdoor-xz-utils-discovered-what-to-know) the keys to pypi for some of my packages that stacks run by various three letter agencies depend on. I've been watching [django-commons](https://github.com/django-commons) with great interest and should participate! This will always be vulnerable to state backed entities playing the long game, but I think the best we can do is figure out systems that translate sweat equity into trust and then require at least 2 eyes on everything that goes out the door.
2. **Robust Continuous Integration** The best rule I have for my own sanity is that I'm not allowed to publish any middle stream open source project without extremely robust CI. "Extremely robust" is squishy so to put some numbers on it: always 100% test coverage, always 100% platform and major dependency minimum and maximum versions in the CI matrix, and usually something like 6x test code to lib code ratio. This front loads the capital investment and makes it much more likely that piling up issues wont tank my resolve going forward. If I can't commit to that level of testing up front then the library just doesn't happen and thats probably for the best. I don't test any of my actual work or personal projects this much. For actualized django websites my CI is usually only robust enough to make sure everything didn't break.
4. **Tool Chain Choice** Thanks to the standardization of pyproject.toml the number of super awesome and useful python build tools has exploded. We all have different preferences but the downside of this choice is the amount of time it takes to figure out a new project's procedural maintenance. You can't tell people what to use so I think the best thing to do is to provide good templates for middle stream projects. I think the tool chain will continue to evolve and help us out here, specifically around automating 1 and 3. I know some maintainers have worked out their own scripting to do this already. We should encourage them to share! Communal maintenance efforts like django-commons may want to consider standardizing all accepted project's toolchains so if you know how to test-build-deploy one of them you have the muscle memory to do it for all of them, and also so the same automation scripts could work for all of them.

**Reply by carltongibson** [reply]
@bckohan I think from what you've said there, I'd take that standardising processes around CI etc, so that maintainers can easily adopt better/best practices would be helpful. 

> … I'm not allowed to publish any middle stream open source project without extremely robust CI. 

That seems like a much higher bar than most projects have. A lot would never even see the light of day. I guess defining **extremely robust** would be helpful. 

FWIW I think it's easy to overplay the trust issues. A lot Django packages aren't in the `xz` type space — form style templates for a particular CSS library, say, aren't generally security sensitive, and are easy to review — but, yes, to the extent they are, you should reserve release rights until you're confident in someone. (That doesn't mean they can't help to get the release ready though; actually tagging and pushing the release, however that's triggered is generally the least of it.)

**Reply by bckohan** [reply]
Yeah, "extremely robust" as I've defined it, is not a standard I hold all third party packages I integrate to. Broad based adoption can inspire as much confidence. I've just found that doing more CI up front reduces my workload in the long run and is the only way sole maintainer projects are feasible for me. If I set something down for a couple years I can still pick it up and do big things with confidence.
